Audit Refactoring #1542 (#1813)
IBM/FHIR: docs/src/pages/guides/FHIRServerUsersGuide.md

  * [4.8 Using local references within request bundles](#48-using-local-references-within-request-bundles)

>   * [4.9 Multi-tenancy](#49-multi-tenancy)

Reviewer: Section 4.9.2.1 (line 1481)
"serviceClassName" : "com.ibm.fhir.audit.logging.impl.DisabledAuditLogService",
Replace with No-op.

Author: hum I don't see it ...

Author: Weird my local copy doesn't show this

Author: OH this is an old commit

Reviewer: ok, thanks!  Not sure why I was seeing that.  Looks good now.

            }
        },
        "audit": {

>             "serviceClassName" : "com.ibm.fhir.audit.logging.impl.DisabledAuditLogService",

Author: Changed here...

Reviewer: Cool.  Thanks!

                    key: binding
                    name: binding-<event_streams_service_instance_name>
```

Please refer to https://cloud.ibm.com/docs/containers?topic=containers-service-binding for detailed instructions if needed.

#### 4.11.3.2 fhir-server-config.json Configuration of audit logging service

```
"audit": {
    "serviceClassName" : "com.ibm.fhir.audit.logging.impl.WhcAuditCadfLogService",
    "serviceProperties" : {
        "load": "config",
        "mapper": "cadf",
        "auditTopic": "FHIR_AUDIT",
        "geoCity": "Dallas",
        "geoState": "TX",
        "geoCounty": "US",
        "kafka" : {
            "sasl.jaas.config": "********",
            "bootstrap.servers": "********",
            "sasl.mechanism": "PLAIN",
            "security.protocol": "SASL_SSL",
            "ssl.protocol": "TLSv1.2",
            "ssl.enabled.protocols": "TLSv1.2",
            "ssl.endpoint.identification.algorithm": "HTTPS"
        },
        "kafkaServers": "********",
        "kafkaApiKey": "********"
    }
```

> The service may be map to the CADF format or the FHIR AuditEvent resource format by declaring a mapper type - 'cadf' or 'auditevent'. 

Reviewer 2: ```suggestion
The service can map to the CADF format or the FHIR AuditEvent resource format by declaring a mapper type - 'cadf' or 'auditevent'. 
```

> The service may be map to the CADF format or the FHIR AuditEvent resource format by declaring a mapper type - 'cadf' or 'auditevent'. 

Reviewer 2: I would have assumed the AuditEvent mapper would just be a different class.  If WhcAuditCadfLogService is used to map to both CADF and FHIR, then I think it should get renamed.

Author: Discussed as a team on code review

        "credential": {
            "token": "user-fhiruser"
        },
        "geolocation": {
            "city": "Dallas",
            "state": "TX",
            "region": "US",
            "annotations": [
            ]
        }
    },
    "observer": {
        "id": "fhir-server",
        "typeURI": "compute/node",
        "name": "IBM FHIR Server - Audit",
        "geolocation": {
            "city": "Dallas",
            "state": "TX",
            "region": "US",
            "annotations": [
            ]
        }
    }
}
```

- *AuditEvent* Example

```
{
    "resourceType": "AuditEvent",
    "type": {
        "system": "http://terminology.hl7.org/CodeSystem/audit-event-type",
        "code": "rest",
        "display": "Restful Operation"
    },
    "subtype": [
        {
            "system": "http://hl7.org/fhir/restful-interaction",
            "code": "search",
            "display": "search"
        }
    ],
    "action": "E",
    "period": {
        "start": "2020-12-10T16:38:14.466Z",
        "end": "2020-12-10T16:38:14.631Z"
    },
    "recorded": "2020-12-10T11:38:14.632173-05:00",
    "outcome": "0",
    "outcomeDesc": "success",

>     "purposeOfEvent": [

Reviewer 2: is there supposed to be an actual purpose string in here somewhere.  having a "purpose" of "PurposeOfUse" seems pretty odd

Reviewer 2: Interesting, apparently that *is* a valid value in the ValueSet.  Its a top-level code which has lots of potential children.  And its the *only* top-level code in the valueset.  So weird.

Author: Yeah - that's why I mentioned that it needs more work once we have a customer asking for a use-case
